Update composed-true-concidered-harmful.md (#1)
Westbrook/blog: composed-true-concidered-harmful.md

These events have two phases; the "capture" phase and the "bubble" phase. During the capture phase, the event travels from the top of the DOM down towards the dispatching element and can be listened for on each of the elements that it passes through in this phase by setting the third argument of `addEventListener()` to true, or by explicitly including `capture: true` in an `options` object passed as the third argument. For example the steps of the "capture" phase of a `click` event on the `<button>` in the following DOM structure:

```html
<body>
    <header>
        <nav>
            <button>Click me!</button>
        </nav>
    </header>
</body>
```

Would be as follows:

1. `<body>`
2. `<header>`
3. `<nav>`
4. `<button>`

Then, being a `click` event has `bubbles: true` by default, the event would enter the "bubble" phase of the event and travel back up the DOM passing through the above DOM in the following order:

1. `<button>`
2. `<nav>`
3. `<header>`
4. `<body>`

> At any point in either phase that you are listening for this event, you will have access to the `preventDefault()`, `stopPropagation()`, and `stopImmediatePropagation()` methods that give you powerful the events that travel across your application. `preventDefault()` can most clearly be felt when listening to a `click` event on an `<a href="...">` tag. In this context, it will _prevent_ the anchor link from being activated and prevent the page from navigating. In a way, this is the event asking for permission to do an action, and we'll look at this more closely in conjunction with manually dispatched events. `stopPropagation()` prevents the event in question from continuing along the DOM tree and triggering listeners along that path, a sort of escape valve for the event when certain parameters are met. This can be taken one step further via `stopImmediatePropagation()` which also prevents the event from completing the current step of the phase it is in. This means that no later bound listeners on that same DOM element for the event in question will be called. So for the `<button>` element in the example above, when a `click` event is dispatched, you could imagine the following completely trivial listeners:

Reviewer: 'that give you powerful the events'?

```suggestion
At any point in either phase that you are listening for this event, you will have access to the `preventDefault()`, `stopPropagation()`, and `stopImmediatePropagation()` methods that give you powerful control over the events that travel across your application. `preventDefault()` can most clearly be felt when listening to a `click` event on an `<a href="...">` tag. In this context, it will _prevent_ the anchor link from being activated and prevent the page from navigating. In a way, this is the event asking for permission to do an action, and we'll look at this more closely in conjunction with manually dispatched events. `stopPropagation()` prevents the event in question from continuing along the DOM tree and triggering listeners along that path, a sort of escape valve for the event when certain parameters are met. This can be taken one step further via `stopImmediatePropagation()` which also prevents the event from completing the current step of the phase it is in. This means that no later bound listeners on that same DOM element for the event in question will be called. So for the `<button>` element in the example above, when a `click` event is dispatched, you could imagine the following completely trivial listeners:
```

Author: `give you powerful control over the events` 🙏

Reviewer: ^ added suggestion for convenience :)

// `stopImmediatePropagation()` method.
header.addEventListener('click', (e) => {
    console.log('not heard on `header` via listener 3 during "bubble"');
});
// You can hear the `click` event during the "capture" phase on the `<button>` element.
button.addEventListener('click', () => {
    coonsole.log('heard on `button` during "capture"');
}, true);

button.click();
// heard on `body` during "capture"
// heard on `button` during "capture"
// heard on `header` via listener 1 during "bubble"
// heard on `header` via listener 2 during "bubble"
```

The majority of values for `bubbles`, `cancelable` (needed to empower `preventDefault()`), and `composed` are the same across native DOM events, and in many of those cases the value of `composed` is `true`, so it's possible that the browser is already refuting the idea that it could "harmful". However, when working with native DOM events the values for these three properties are also not configurable. To access the power, and responsibility, that comes with being able to do so, you'll need to enter the world of manually dispatched events.

## `dispatchEvent()` <a id="manual-dispatch"></a>

> So far we've mainly talked about the `click` event as automatically dispatched by the browser. There is, of course, a whole family of UA-dispatched UI events that can be addressed in the same manner (e.g. `animationend`/`copy`/`keydown`/`mouseover`/`paste`/`touch`, etc.). However, the real fun starts when you take that power into your own hands and start dispatching events on your own creation. For this, the browser supplies us with the [`dispatchEvent()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent) method that hangs off of anything extended from `EventTarget`, which includes all of the `HTMLElement` based class of DOM elements. For this to do its magic we need to supply it an event to dispatch. We're given a number of events classes to creat our new event from (e.g. `new Event()`, `new MouseEvent()`, `new InputEvent()`, etc.), but for the most part working with `new Event(typeArg[, initDict])` only is more than enough. 

Reviewer: ```suggestion
So far we've mainly talked about the `click` event as automatically dispatched by the browser. There is, of course, a whole family of UA-dispatched UI events that can be addressed in the same manner (e.g. `animationend`/`copy`/`keydown`/`mouseover`/`paste`/`touch`, etc.). However, the real fun starts when you take that power into your own hands and start dispatching events on your own creation. For this, the browser supplies us with the [`dispatchEvent()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent) method that hangs off of anything extended from `EventTarget`, which includes all of the `HTMLElement` based class of DOM elements. For this to do its magic we need to supply it an event to dispatch. We're given a number of events classes to create our new event from (e.g. `new Event()`, `new MouseEvent()`, `new InputEvent()`, etc.), but for the most part working with `new Event(typeArg[, initDict])` only is more than enough. 
```

Now, we're ready to dispatch an event.

```js
el.dispatchEvent('test-event');
```

Event dispatched!

The event has a `type` of `test-event`, so a listener like the following will be able to hear it:

```js
el.addEventListener('test-event', (e) => console.log(e.type));
// test-event
```

You can also listen for this event during the "capture" phase:

```js
const body = document.querySelector('body');
body.addEventListener('test-event', (e) => console.log(e.type), true);
// test-event
```

But, you won't be hearing it in the "bubble" phase:

```js
const body = document.querySelector('body');
body.addEventListener('test-event', (e) => console.log(e.type));
// ... ... Bueller?
```

This is because by default a `new Event()` (as well as all derivative event constructors) have `bubbles`, `cancelable`, and `composed` set to `false` by default. This is where the optional `initDict` argument of our event constructor comes into play. When you want to customize the values of these, you'll create your event like so:

```js
const event = new Event('test-event', {
    bubbles: true,
    cancelable: true,
    composed: true,
};
```

> Or however best supports (or least harms? 😉) the use case in question. That means that if you only want your event to be available in the "capture" phase (which literally means it takes half the time for it to run synchronously through your application that it would if it were to also make a pass through the "bubble" phase) you can leave that out. Don't have an action that you'd like permission to do? You can leave out `cancelable`, too. Don't have shadow DOM? Decided definatively that `composed: true` is harmful? It's your rodeo, leave it out!

Reviewer: ```suggestion
Or however best supports (or least harms? 😉) the use case in question. That means that if you only want your event to be available in the "capture" phase (which literally means it takes half the time for it to run synchronously through your application than it would if it were to also make a pass through the "bubble" phase) you can leave that out. Don't have an action that you'd like permission to do? You can leave out `cancelable`, too. Don't have shadow DOM? Decided definitely that `composed: true` is harmful? It's your rodeo, leave it out!
```

When working with nested shadow roots (e.g. custom elements with custom elements inside of them) this event retargeting will happen at each shadow boundary that the event encounters. That means that if there are three shadow roots that the event passed through the `target` will change three times:

```html
<body> <-- target: parent-el -->
    <parent-el> <-- target: parent-el -->
        #shadow-root <-- target: child-el -->
            <child-el> <-- target: child-el -->
                #shadow-root <-- target: grandchild-el -->
                    <grandchild-el> <-- target: grandchild-el -->
                        #shadow-root <-- target: button -->
                            <button> <-- target: button -->
                                Click here!
                            </button> <!-- click happens here -->
                    <grandchild-el>
            <child-el>
    <parent-el>
</body>
```

This is, of course, one of the benefits of the encapsulation that a shadow root can provide, what happens in the shadow root stays in the shadow root and beyond the shadow root, a less direct version of those actions are available.

### The Composed Path Less Traveled

> There are times when we need a look into that dirty laundry to get a peek at just where that event came from, be it `<button>`, `<div>`, `<a>`, or something else (it's hopefully a `<button>` or `<a>`...a11y people!), and for those times we've got the `composedPath()` method on our events. At any point in the event's lifecycle, calling `composedPath()` on that event will give you an array of all the DOM elements on which it will be heard. The array is listed in "bubble" order, so the zeroeth item will be the dispatching element and the last item will be the last element through which the event will pass. That means you can always use the following code to ascertain the original dispatching element and outline the path along which the event will trave, assuming the previous example HTML:

Reviewer: ```suggestion
There are times when we need a look into that dirty laundry to get a peek at just where that event came from, be it `<button>`, `<div>`, `<a>`, or something else (it's hopefully a `<button>` or `<a>`...a11y, people!), and for those times we've got the `composedPath()` method on our events. At any point in the event's lifecycle, calling `composedPath()` on that event will give you an array of all the DOM elements on which it will be heard. The array is listed in "bubble" order, so the zeroeth item will be the dispatching element and the last item will be the last element through which the event will pass. That means you can always use the following code to ascertain the original dispatching element and outline the path along which the event will trave, assuming the previous example HTML:
```

```html
<div>
    <shadow-root-el>
        #shadow-root
            <section>
                <div>
                    <event-dispatching-element>
```

It's really `composed: false` that gives us new and interesting functionality.

> When an event is disaptch with `composed: false` then that event will be contained with the shadow root in which it is fired. Righ off, for the speed-obsessed developers reading this, that means your events will go faster! Whereas `{bubbles: false}` can double the speed of an event by completely cutting off the "bubble" phase (read half of the traveling required of an event), `{composed: false}` could cut that distance all the way down to two stops, the dispatching element and the shadow root that contains it, assuming such a simplified DOM tree. Code speed is likely not the concern here, but it is worth noting. What's really of most interest is access. When an event is dispatched with `composed: false` only the other elements encapsulated in its shadow root have access to it.

Reviewer: ```suggestion
When an event is dispatch with `composed: false` then that event will be contained within the shadow root in which it is fired. Right off, for the speed-obsessed developers reading this, that means your events will go faster! Whereas `{bubbles: false}` can double the speed of an event by completely cutting off the "bubble" phase (read half of the traveling required of an event), `{composed: false}` could cut that distance all the way down to two stops, the dispatching element and the shadow root that contains it, assuming such a simplified DOM tree. Code speed is likely not the concern here, but it is worth noting. What's really of most interest is access. When an event is dispatched with `composed: false` only the other elements encapsulated in its shadow root have access to it.
```
